getspice.py

```diff
@@ -43,114 +41,212 @@
+
+            # Download the file
+            print('Downloading CSPICE for {0}...'.format(distribution))
+            attempts = 10   # Let's try a maximum of attempts for getting SPICE
+            while attempts:
+                attempts -= 1
+                try:
+                    self._download()
+                except RuntimeError as error:
+                    print("Download failed with URLError: {0}, trying again after "
+                          "15 seconds!".format(error))
+                else:
+
+                    # Unpack the file
+                    print('Unpacking... (this may take some time!)')
+                    self._unpack()
+
+                # We are done.  Let's return to the calling code.
+                break
+
+    def _distribution_info(self):
+        """Creates the distribution name and the expected extension for the
+        CSPICE package and returns it.
+
+        :return (distribution, extension) tuple where distribution is the best
+                guess from the strings available within the platform_urls list
+                of strings, and extension is either "zip" or "tar.Z" depending
+                on whether we are dealing with a Windows platform or else.
+        :rtype: tuple (str, str)
+
+        :raises: KeyError if the (system, machine) tuple does not correspond
+                 to any of the supported SpiceyPy environments.
+        """
+
+        print('Gathering information...')
+        system = platform.system()
+
+        # Cygwin system is CYGWIN-NT-xxx.
+        system = 'cygwin' if 'CYGWIN' in system else system
+
+        processor = platform.processor()
+        machine = '64bit' if sys.maxsize > 2 ** 32 else '32bit'
+
+        print('SYSTEM:   ', system)
+        print('PROCESSOR:', processor)
+        print('MACHINE:  ', machine)
+
+        return self._dists[(system, machine)]
+
+    def _download(self):
+        """Support function that encapsulates the OpenSSL transfer of the CSPICE
+        package to the self._local io.ByteIO stream.
+
+        :raises RuntimeError if there has been any issue with the HTTPS
+                             communication
+
+        .. note::
+
+           Handling of CSPICE downloads from HTTPS
+           ---------------------------------------
+           Some Python distributions may be linked to an old version of OpenSSL
+           which will not let you connect to NAIF server due to recent SSL cert
+           upgrades on the JPL servers.  Moreover, versions older than
+           OpenSSL 1.0.1g are known to contain the 'the Heartbleed Bug'.
+           Therefore this method provides two different implementations for the
+           HTTPS GET call to the NAIF server to download the required CSPICE
+           distribution package.
+        """
+        # Use urllib3 (based on PyOpenSSL).
+        if ssl.OPENSSL_VERSION < 'OpenSSL 1.0.1g':
+            import pip
+            pip.main(['install', 'pyOpenSSL'])
```

Comment:
ok, so when I first saw this, I was not a huge fan of runtime invocations of pip like this. But I thought of an alternative solution that may be worth trying. How about we just have a runtime check (the if statement above this line) and have it update the `DEPENDENCIES` list to include urllib3 and pyOpenSSL? This way, we would not need to import pip and the runtime installations could be removed. ie at the top of setup.py have a code block like: 
```python
DEPENDENCIES = ['numpy>=1.8.0', 'six>=1.9.0']
if ssl.OPENSSL_VERSION < 'OpenSSL 1.0.1g':
    DEPENDENCIES.append('pyOpenSSL>=17.3.0')
    DEPENDENCIES.append('urllib3[secure]>=1.22')
```